compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/PipelineTranslator.java

```diff
@@ -198,11 +196,29 @@ private static void parDoMultiOutputTranslator(final TranslationContext ctx,
         pValueWithTupleTag.getKey()));
   }
 
+  private static GroupByKeyTransform createGBKTransform(
+    final TranslationContext ctx,
+    final TransformVertex transformVertex) {
+    final AppliedPTransform pTransform = transformVertex.getNode().toAppliedPTransform(PIPELINE.get());
+    final PCollection<?> mainInput = (PCollection<?>)
+      Iterables.getOnlyElement(TransformInputs.nonAdditionalInputs(pTransform));
+    final TupleTag mainOutputTag = new TupleTag<>("main output");
+
+    return new GroupByKeyTransform(
+      getOutputCoders(pTransform),
+      mainOutputTag,
+      Collections.emptyList(),
+      mainInput.getWindowingStrategy(),
+      Collections.emptyList(), /* side inputs */
+      ctx.pipelineOptions,
+      SystemReduceFn.buffering(mainInput.getCoder()));
```

Comment:
I skimmed through `SystemReduceFn`, and noticed that we have two options here.
- `SystemReduceFn.buffering()`: Buffers all input values in persistent state and produces an iterable over it
- `SystemReduceFn.combining()`: Combines all input values using a `CombineFn`

My guess is that `SystemReduceFn.combining()` provides better performance, assuming that the operation is combinable and a data reduction happens with combining. Maybe we should try to use `SystemReduceFn.combining()` as much as we can?

Also in both cases a buffer with a state id is created and used. Does that mean our `InMemoryStateInternalsFactory` will be used by `SystemReduceFn`? It'd be nice to leave a comment(TODO) regarding this as we may want to use different types of stores for `SystemReduceFn` operations.

Reply:
Nemo currently does not perform incremental combining. 
It groups inputs according to the window and key and combine the collected inputs after grouping is done for all bounded inputs. Using `System.ReduceFn.combining` can enable incremental combining, and I think this is a separate issue. I've filed an issue for this https://issues.apache.org/jira/browse/NEMO-222